src/DataTypes/DataTypeFactory.cpp

```diff
@@ -76,6 +78,13 @@ DataTypePtr DataTypeFactory::get(const String & family_name_param, const ASTPtr
         return get("LowCardinality", low_cardinality_params);
     }
 
+    if (CurrentThread::isInitialized())
+    {
+        auto query_context = CurrentThread::get().getQueryContext();
+        if (query_context && query_context->getSettingsRef().log_queries)
```

Comment:
The settings are shared between threads, let's see if TSan will find something.
If so, we can copy the log_queries field to the local variable inside ThreadStatus.